File: docs/examples/todo/node/src/repositories/TodoRepository.js
/**
 * Todo repository interface.
 *
 * Defines the contract for todo data access. Concrete implementations
 * can use in-memory storage, a database, or any other persistence layer.
 *
 * @interface
 */
class TodoRepository {
  /**
   * Persist a new todo.
   *
   * @param {{ title: string, description?: string }} data - Todo creation payload.
   * @returns {Promise<{ id: string, title: string, description: string, done: boolean, createdAt: string }>}
   */
  async save(data) {
    throw new Error('Not implemented');
  }

  /**
   * Retrieve all todos.
   *
   * @returns {Promise<Array<{ id: string, title: string, description: string, done: boolean, createdAt: string }>>}
   */
  async findAll() {
    throw new Error('Not implemented');
  }

  /**
   * Mark an existing todo as done.
   *
   * @param {string} id - Todo identifier.
   * @returns {Promise<{ id: string, title: string, description: string, done: boolean, createdAt: string } | null>}
   */
  async markDone(id) {
    throw new Error('Not implemented');
  }
}

module.exports = TodoRepository;

File: docs/examples/todo/python/app/repositories/todo_repository.py
"""Abstract base class for todo data access (Dependency Inversion Principle)."""

from abc import ABC, abstractmethod
from typing import Optional


class TodoRepository(ABC):
    """Repository interface for todo data operations.

    Concrete implementations provide the actual storage mechanism.
    Higher-level modules depend on this abstraction, not on a specific database.
    """

    @abstractmethod
    def save(self, title: str, description: str = "") -> dict:
        """Persist a new todo.

        Args:
            title: Required todo title.
            description: Optional description text.

        Returns:
            Dict with keys: id, title, description, done, created_at.
        """

    @abstractmethod
    def find_all(self) -> list[dict]:
        """Retrieve all todos.

        Returns:
            List of todo dicts. Empty list when none exist.
        """

    @abstractmethod
    def mark_done(self, todo_id: str) -> Optional[dict]:
        """Mark an existing todo as complete.

        Args:
            todo_id: Unique todo identifier.

        Returns:
            Updated todo dict, or None if not found.
        """
